### DISCUSSION TITLE
create a concrete value during the tracing

### DISCUSSION THREAD

**Discussion by mariogeiger** [jax-ml/jax#7767] [Q&A] [answered]
Hi,

Is there a way to create concrete arrays while being in a tracer call?
```python
@jit
def f():
    x = jnp.array([1.0, 2.0])
    # I want a concrete x because I want to use it only during the tracing
```

Here is why I need that. Let say I have two cases whether a function is symmetric or antisymmetric and I have a function to check that numerically:
```python
def parity(f):
    a = jnp.linspace(-1.0, 1.0, 21)
    if jnp.allclose(f(a), f(-a)):
        return 'symmetric'
    if jnp.allclose(f(a), -f(-a)):
        return 'antisymmetric'
    return 'undefined'
```

The problem is that if I use `parity` inside a function that gets `jit`ed I get an error because `a` is not concrete:
```
ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected: Traced<ShapedArray(bool[])>with<DynamicJaxprTrace(level=0/1)>
The problem arose with the `bool` function. 
```

How to force `a` to be concrete?

**Comment by mariogeiger** [comment]
I dream of something like a context manager
```python
def parity(f):
    with jax.concrete():
        a = jnp.linspace(-1.0, 1.0, 21)
        if jnp.allclose(f(a), f(-a)):
            return 'symmetric'
        if jnp.allclose(f(a), -f(-a)):
            return 'antisymmetric'
        return 'undefined'
```

**Comment by jakevdp** [comment, marked as answer]
The easiest way to make a value (or an operation) concrete, is to use the `np` version rather than the `jnp` version:
```python
def parity(f):
    a = np.linspace(-1.0, 1.0, 21)
    if np.allclose(f(a), f(-a)):
        return 'symmetric'
    if np.allclose(f(a), -f(-a)):
        return 'antisymmetric'
    return 'undefined'
```

**Reply by mariogeiger** [reply]
Thanks, Yes I thought too and I tried it but it does not work if `f` is a jax function, let say `jax.nn.gelu`

**Reply by jakevdp** [reply]
To be clear, this works for arbitrary functions so long as the outer function is not traced. e.g.
```python
def parity(f):
    a = np.linspace(-1.0, 1.0, 21)
    if np.allclose(f(a), f(-a)):
        return 'symmetric'
    if np.allclose(f(a), -f(-a)):
        return 'antisymmetric'
    return 'undefined'

parity(jax.nn.gelu)
# "undefined"
```
But from your response, it sounds like the outer function *is* traced, and you're not concerned about concrete values so much as forcing static computation for certain blocks within a traced function. I don't think JAX has any mechanism for that.

**Reply by mariogeiger** [reply]
Yes exactly, I want to force a block of code to be statically computed.
Would it be something interesting to add to the library?

**Reply by soraros** [reply]
What about something like this

```python
def parity(f):
  a = jnp.linspace(-1.0, 1.0, 21)
  if jnp.allclose(f(a), f(-a)):
    return 'symmetric'
  elif jnp.allclose(f(a), -f(-a)):
    return 'antisymmetric'
  else:
    return 'undefined'

def h(f, x):
  with jax.core.eval_context():
    p = parity(f)
    n = {'s': 1, 'a': -1, 'u': 0}[p[0]]
    print(p)
  return n * f(x), n

jax.make_jaxpr(partial(h, jnp.sin))(jnp.arange(10.))
# antisymmetric
# { lambda  ; a.
#   let b = sin a
#       c = mul b -1.0
#   in (c, -1) }
```

**Reply by mariogeiger** [reply]
Amazing !

**Reply by jakevdp** [reply]
I don't think `core.eval_context` is considered a public API, so if you use it be prepared for it to change without warning in future releases.

**Reply by mariogeiger** [reply]
Of course

**Reply by bonfab** [reply]
I have similar problem where this would also be the most elegant solution. Do plans exist to make a jax context manager officially available?

**Reply by soraros** [reply]
FYI, there is #7535
